Fix bug where list fields don't report their logical type correctly

jhump opened the pull request:
When loading a parquet file, the LIST logical type is never present, even if it's present in the actual schema in the file. This fixes that.

achille-roussel reviewed (commented):
Thank you so much for the fix!

Would you be able to contribute a simple test that ensures that the new code path is exercised and that we won't introduce regressions?

jhump commented:
@achille-roussel, I couldn't find any tests similar to what you're asking for. So I back-filled a category of tests for making sure the schema survives a round-trip of being written to a file and then read back.

It's kind of big, so apologies if it's not appropriate or not what you had in mind. If that's the case, please provide a little more direction and maybe I can come up with what you're hoping to see.

FWIW, this test does tickle this bug. Without the two-line change in `column.go`, there is a test failure in the new test:
```
=== RUN   TestSchemaRoundTrip/lists
    schema_test.go:519: 
        expected:
        
        message root {
        	optional group lists {
        		required group groups (LIST) {
        			repeated group list {
        				optional group element {
        					required binary a (STRING);
        					required int32 b (INT(32,true));
        				}
        			}
        		}
        		required group ints (LIST) {
        			repeated group list {
        				required int32 element (INT(32,true));
        			}
        		}
        		optional group strings (LIST) {
        			repeated group list {
        				required binary element (STRING);
        			}
        		}
        	}
        }
        
        found:
        
        message root {
        	optional group lists {
        		required group groups {
        			repeated group list {
        				optional group element {
        					required binary a (STRING);
        					required int32 b (INT(32,true));
        				}
        			}
        		}
        		required group ints {
        			repeated group list {
        				required int32 element (INT(32,true));
        			}
        		}
        		optional group strings {
        			repeated group list {
        				required binary element (STRING);
        			}
        		}
        	}
        }
--- FAIL: TestSchemaRoundTrip/lists (0.00s)
```

The test is kind of big since it seemed useful to also exhaustively verify all of the other supported logical types. Note that I've added some TODOs to the tests because there are a few things that parquet-go does not seem to support though they are in the spec:
1. Decimal logical types can be backed by byte array columns. But parquet-go only accepts that logical type with int32, int64, and fixed len byte array columns.
2. Times and timetamps support "local" times, where the value is not normalized to UTC. Under the hood, parquet-go appears to work with this, but it provides no exported API for actually constructing such a field.

jhump commented:
@achille-roussel, ping.

jhump commented:
@achille-roussel, ping.

jhump commented:
@achille-roussel, any chance you'll have time to look at this soon? Please let me know if there's anything you want me to change.

jhump commented:
The failing test appears to be due to one of the test files being non-compliant: it uses a 3-level logical list, but the leaf field is named "item" instead of "element". Is it possible these files were created with a version of pyarrow affected by this bug: https://issues.apache.org/jira/browse/ARROW-11497?

I could either fix the test data file, to change the field name to "element", or I could make the existing check lenient and have it allow either "item" or "element". Do you have an opinion on how I should proceed?

jhump commented:
FWIW, I pushed a change to make it lenient. All tests pass locally. (The CI run also had some buffer pool release issues, but I can't repro those and don't think they can possibly be related to my patch.)

Please take another look. If you'd prefer I leave the check stricter and instead fix the test file with the "item" field name, let me know.

achille-roussel reviewed (approved):
Amazing, thanks for all the tests you added!
